### PR TITLE
Cache subscriptions

### PR DESCRIPTION
**Pull request by SimonCropp**
This PR introduces caching subscriptions for a specified amount of time. This reduces latency by skipping an Azure Storage Table lookup when publishing messages.

### CODE REVIEW

**Comment by SimonCropp** [comment]
@Particular/azure-maintainers this is ready for review

File: src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs [hunk 1]
```diff
@@ -91,15 +89,65 @@ static string DecodeFrom64(string encodedData)
...
+                valueFactory: _ => new CacheItem
+                {
+                    Stored = DateTime.UtcNow,
+                    Subscribers = GetSubscriptions(types)
+                });
+
+            var age = DateTime.UtcNow - cacheItem.Stored;
+            if (age >= cacheFor)
+            {
+                cacheItem.Subscribers = GetSubscriptions(types);
```
Comment (Scooletz) [inline src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs]: This updates an item that is stored in a concurrent dictionary, which is not atomically visible for readers. I know that we bump up the `Stored` after updating subscribers, but I think, that being explicit and accessing the Cache again, registering a new item if nobody updated it meanwhile is a better approach.

File: src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs [hunk 2]
```diff
@@ -142,5 +190,11 @@ public int GetHashCode(Subscriber obj)
                 return (obj.Endpoint ?? string.Empty).Length;
             }
         }
+
+        internal class CacheItem
+        {
+            public DateTime Stored;
+            public Task<IEnumerable<Subscriber>> Subscribers;
```
Comment (Scooletz) [inline src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs]: A question. Would storing awaited result be cheaper? In other words: the task that is returned from `GetSubscriptions` could have something captured in lambda/state machine. I'd assume that if we stored `Task.FromResult (await GetSubscriptions)` this could introduce less overhead. Am missing something?

File: src/NServiceBus.Persistence.AzureStorage.ComponentTests/packages.config [hunk 3]
```diff
@@ -8,9 +8,10 @@
   <package id="Microsoft.Data.OData" version="5.8.2" targetFramework="net452" />
   <package id="Microsoft.Data.Services.Client" version="5.8.2" targetFramework="net452" />
   <package id="Mono.Cecil" version="0.9.6.4" targetFramework="net452" />
-  <package id="Newtonsoft.Json" version="9.0.1" targetFramework="net452" />
+  <package id="Newtonsoft.Json" version="10.0.1" targetFramework="net452" />
```
Comment (Scooletz) [inline src/NServiceBus.Persistence.AzureStorage.ComponentTests/packages.config]: I'd keep bumping up dependency as a separate PR/commit.

File: src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs (hunk 2 above)
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs]: >  Would storing awaited result be cheaper? 

You mean storing tasks. Yes, I think that would be effective. You don't need to store `Task.FromResult(await GetSub())`, just a task returned by the call. If task is awaited, then it's completed. And completed task`s result can be used multiple times.

Comment (Scooletz) [inline src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs]: Nope. What I meant is that behind a task there can be a state machine/variables captured in a closure. I'm asking if awaiting the result and then re-wrapping it in a new Task is a better/cheaper approach, that holds less objects.
☝️ this might an over-engineering though.

Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs]: Personally, I find the last comment to add more complexity than clarity.
Get a task, cache it. Await for the result. Next time cached task is awaited, it's going to return the straight result.

Comment (Scooletz) [inline src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs]: I think that I can't rephrase it in a better way. Leave it as it is. It's not that important.

**Comment by Scooletz** [comment]
Ping @SimonCropp I've added my comments.

File: src/NServiceBus.Persistence.AzureStorage.ComponentTests/packages.config (hunk 3 above)
Comment (SimonCropp) [inline src/NServiceBus.Persistence.AzureStorage.ComponentTests/packages.config]: This was required since i used objectapproval to do the verification

File: src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs (hunk 2 above)
Comment (SimonCropp) [inline src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs]: @danielmarbach thoughts?

File: src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs (hunk 1 above)
Comment (SimonCropp) [inline src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs]: Wouldnt a comment suffice?

File: src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs (hunk 2 above)
Comment (SimonCropp) [inline src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs]: so i used this approach for several reasons.

 * Avoids an async state machine for `GetSubscriberAddressesForMessage`
 * Allows us to always return from this method in a non blocking way
 * defers the decision to await/or not to the caller
 * simplifies the code

File: src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs (hunk 1 above)
Comment (danielmarbach) [inline src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs]: I agree with @Scooletz using the update APIs of the concurrent dictionary might be a more clear approach

File: src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs (hunk 2 above)
Comment (danielmarbach) [inline src/NServiceBus.Persistence.AzureStorage/Subscriptions/AzureSubscriptionStorage.cs]: What @SimonCropp summarized. In this case there is no state machine involved and therefore no allocs beside the task. In the future this might be a good opportunity to use a ValueTask.

**Comment by SimonCropp** [comment]
@Particular/azure-maintainers i think i am done here

**Comment by SimonCropp** [comment]
please let me know if you want documentation written

**Comment by Scooletz** [comment]
I'm merging this.  
The new config extension method should be documented
```
public static PersistenceExtensions<AzureStoragePersistence, StorageType.Subscriptions>CacheFor
     (this PersistenceExtensions<AzureStoragePersistence, StorageType.Subscriptions> config, 
     TimeSpan timeSpan)
```

probably [here](https://docs.particular.net/nservicebus/azure-storage-persistence/configuration).

**Comment by SimonCropp** [comment]
Did u want me to do a pr for the doco

**Comment by Scooletz** [comment]
Yes @SimonCropp
